pkg/autoscaler/scaling/autoscaler.go

```diff
@@ -169,7 +169,13 @@ func (a *autoscaler) Scale(logger *zap.SugaredLogger, now time.Time) ScaleResult
 		} else {
 			logger.Errorw("Failed to obtain metrics", zap.Error(err))
 		}
-		return invalidSR
+		if errors.Is(err, metrics.ErrNoData) && !a.panicTime.IsZero() && a.panicTime.Add(spec.StableWindow).Before(now) {
+			// We are not receiving data, but we panicked before, and the panic stable window has gone, we want to continue.
+			// We want to un-panic and scale down to 0 if the rest of the conditions allow it
```

Comment:
> when requests timed out, metrics will be reported. As there are not requests pending, the autoscaler would want to scale down to 0, but because we are still in panic mode, this will be ignored and the desiredPanicPodCount will be applied instead.

what is the value of `desiredPanicPodCount` ? I'm wondering in this scenario if we should opt of panic'ing

Reply:
hard to say, it depends on the calculations done in the function. In line 205 you can see this
```
	// We want to keep desired pod count in the  [maxScaleDown, maxScaleUp] range.
	desiredStablePodCount := int32(math.Min(math.Max(dspc, maxScaleDown), maxScaleUp))
	desiredPanicPodCount := int32(math.Min(math.Max(dppc, maxScaleDown), maxScaleUp))
```

`dppc` is calculated like (line 194)
```
	dspc := math.Ceil(observedStableValue / spec.TargetValue)
	dppc := math.Ceil(observedPanicValue / spec.TargetValue)

```

and these values comes from the metricClient (line 163)
```
observedStableValue, observedPanicValue, err = a.metricClient.StableAndPanicConcurrency(metricKey, now)
```

Reply:
I run a manual test again.
I setup a ksvc revision with `containerConcurrency: 2`. This is a failing revision, it will not start up.
This revision also has `timeoutSeconds: 30`, which means that request at the activator will timeout after 30 seconds.

I created 4 concurrent requests, which made the autoscaler to determine that 3 pods were needed.

For some seconds these were the values for 
```
desiredStablePodCount = 3, desiredPanicPodCount = 3
```
I saw as well the message `Operating in panic mode.` during this period of time.

[This block of code](https://github.com/knative/serving/blob/371744649dbce6aa3086d178f8fe08fefffce813/pkg/autoscaler/scaling/autoscaler.go#L230-L246) makes sure that `a.maxPanicPods` is the greater of desiredStablePodCount and desiredPanicPodCount. Which at this point is `3`.

At some point the values changed to
```
desiredStablePodCount = 3, desiredPanicPodCount = 2
```
and finally to
```
desiredStablePodCount = 3, desiredPanicPodCount = 0
```
then it moved to 
```
desiredStablePodCount = 2, desiredPanicPodCount = 0
```
but this decision is skipped due to 
`Skipping pod count decrease from 3 to 2.` (see condition in the above code)

then, even though the values are:
```
desiredStablePodCount = 0, desiredPanicPodCount = 0
```
I still see the skipping `Skipping pod count decrease from 3 to 0.` Because we are in `panic mode` and we do not decrease values when we are in panic mode.
We are getting to `desiredStablePodCount = 0` because there are no more requests at the activator (they timed out)

Without the changes in this PR. No more metrics will be received from now on for this revision. [And this block of code](https://github.com/knative/serving/blob/371744649dbce6aa3086d178f8fe08fefffce813/pkg/autoscaler/scaling/autoscaler.go#L166-L173) will always be true. 
```
	if err != nil {
		if errors.Is(err, metrics.ErrNoData) {
			logger.Debug("No data to scale on yet")
		} else {
			logger.Errorw("Failed to obtain metrics", zap.Error(err))
		}
		return invalidSR
	}

```
And the revision will remain in `panic mode` from now on.

The new condition aims for allowing another execution of the function if the revision panicked and the stable window has gone so the revision will scale down.

Reply:
I still think this change is needed. Let me know your thoughts about the above explanation.